### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Sites): the category of sheaves is a localization of the category of presheaves

### PR DESCRIPTION
**Pull request by joelriou**
In this PR, it is shown that the category of sheaves `Sheaf J A` is a localization of the category `Presheaf J A` with respect to the class `J.W` of morphisms of presheaves which become isomorphisms after applying the sheafification functor.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/Localization.lean
Comment (erdOne) [inline Mathlib/CategoryTheory/Sites/Localization.lean:10]:
```suggestion
/-!
# The sheaf category as a localized category

```

**Review by erdOne — COMMENTED** [review]
I think the whole file holds for general reflective categories? We might not need this generality though.

Anyways LGTM modulo the minor comment.

**Comment by erdOne** [comment]
Also is `W` standard nomenclature?

**Comment by joelriou** [comment]
> I think the whole file holds for general reflective categories? We might not need this generality though.

Thanks for the review. I have shown the general localization lemma for an adjunction for which the right adjoint is fully faithful, but here I was able to define the class of morphisms `W` and some properties without even assuming that the adjoint exists! (In the applications, we shall use two different left adjoints to the forget functor from sheaves to presheaves, the abstract one given by `HasWeakSheafify`, and the one given by the plus-plus-construction.) Then, having a specialized API seems worth to me.

I have used `W` because it is the letter I have used throughout the localization of categories API. It appears it is used as well in https://ncatlab.org/nlab/show/category+of+sheaves

**Comment by erdOne** [comment]
Ah so I guess it stands for weak equivalences
For general fully-faithful embeddings, I think
```
fun _ P₂ f => ∀ Q, Function.Bijective (fun (g : P₂ ⟶ F.obj Q) => f ≫ g)
```
works?

**Comment by joelriou** [comment]
I think your definition works more generally. In order to avoid duplication of code, I think the most important is that I first develop classes of morphisms with the 2-out-of-3 property...

**Comment by joelriou** [comment]
Indeed, your suggestion is very much related to definitions in the context of left Bousfield localization of model categories. I have included very basic API for left Bousfield localization, and deduce that sheafification is a localization.

File: Mathlib/CategoryTheory/Localization/Bousfield.lean
Comment (jcommelin) [inline Mathlib/CategoryTheory/Localization/Bousfield.lean:47]:
```suggestion
/-- Given a predicate `P : C → Prop`, this is the class of morphisms `f : X ⟶ Y`
```

**Comment by jcommelin** [comment]
@erdOne If you are happy with this, please go ahead and merge it.

bors d=erdOne

**Review by erdOne — APPROVED** [review]
Small nitpick. Otherwise LGTM.

File: Mathlib/CategoryTheory/Localization/Bousfield.lean
Comment (erdOne) [inline Mathlib/CategoryTheory/Localization/Bousfield.lean:104]: This is abusing the `Set` defeq. Maybe `(· ∈ Set.range F.obj)` instead?

**Comment by erdOne** [comment]
Thanks! LGTM
Feel free to merge if you are happy with the PR (This is still awaiting-author so I did not merge it).

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**
